File: api/app/core/money.py
"""EUR-pivot currency conversion for the monthly Telegram report — ported
from server.js's convertAmount/convertToEurAmount/getExpenseForMonth.
"""

from __future__ import annotations

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app import models


def convert_to_eur_amount(amount, currency: str, meta: dict) -> float:
    value = float(amount or 0)
    if currency == "RUB":
        return value / (float(meta.get("rateRubPerEur")) or 100)
    if currency == "USDT":
        return value / (float(meta.get("rateUsdtPerEur")) or 1.08)
    return value


def convert_amount(amount, from_currency: str, to_currency: str, meta: dict) -> float:
    from_currency = from_currency or "USDT"
    if from_currency == to_currency:
        return float(amount or 0)
    eur = convert_to_eur_amount(amount, from_currency, meta)
    if to_currency == "EUR":
        return eur
    if to_currency == "RUB":
        return eur * (float(meta.get("rateRubPerEur")) or 100)
    if to_currency == "USDT":
        return eur * (float(meta.get("rateUsdtPerEur")) or 1.08)
    return eur


def get_expense_for_month(db: Session, month_key: str, meta: dict) -> float:
    rows = db.execute(
        select(models.Payment.amount, models.Payment.currency).where(
            func.substr(models.Payment.paid_at, 1, 7) == month_key
        )
    ).all()
    return sum(convert_amount(row.amount, row.currency, "RUB", meta) for row in rows)

File: web/lib/money.js
"use client";

import { useAuth } from "../context/AuthContext";
import { useLocale } from "../context/LocaleContext";

// Ported verbatim from App.vue's convertToEur/convertAmount/usdRubRate/
// formatMoney/paymentsTotalIn/formatPaymentTotal/totalPayments — EUR stays the
// backend's pivot currency (see CLAUDE.md). Parameterized by meta/locale
// instead of reading `this.*` so they're plain, testable functions; useMoney()
// below binds them to context the way call sites actually use them.

export function convertToEur(amount, currency, meta) {
  const value = Number(amount || 0);
  if (currency === "RUB") return value / (Number(meta.rateRubPerEur) || 100);
  if (currency === "USDT") return value / (Number(meta.rateUsdtPerEur) || 1.08);
  return value;
}

export function convertAmount(amount, from, to, meta) {
  const fromCurrency = from || "USDT";
  const toCurrency = to || meta.currency || "USDT";
  if (fromCurrency === toCurrency) return Number(amount || 0);
  const eur = convertToEur(amount, fromCurrency, meta);
  if (toCurrency === "EUR") return eur;
  if (toCurrency === "RUB") return eur * (Number(meta.rateRubPerEur) || 100);
  if (toCurrency === "USDT") return eur * (Number(meta.rateUsdtPerEur) || 1.08);
  return eur;
}

export function usdRubRate(meta) {
  return Number(meta.rateRubPerEur || 0) / (Number(meta.rateUsdtPerEur) || 1);
}

// For currency *pickers* (option labels), not amounts — formatMoney already
// renders EUR/RUB amounts with their Intl-native €/₽ symbols; this is what
// those selects need to show the same symbols instead of full names like
// "Российский рубль", which wrapped to 2 lines and forced the select (and
// any grid/flex track sizing it) wider than intended.
export function currencySymbol(currency) {
  return { USDT: "₮", EUR: "€", RUB: "₽" }[currency] || currency;
}

export function formatMoney(value, currency = "USDT", locale = "ru") {
  const intlLocale = locale === "en" ? "en-US" : "ru-RU";
  const num = Number(value || 0);
  if (currency === "EUR" || currency === "RUB") {
    return new Intl.NumberFormat(intlLocale, { style: "currency", currency, maximumFractionDigits: 2 }).format(num);
  }
  return `${new Intl.NumberFormat(intlLocale, { minimumFractionDigits: 2, maximumFractionDigits: 2 }).format(num)} ₮`;
}

export function paymentsTotalIn(payments = [], targetCurrency, meta) {
  return (payments || []).reduce((sum, payment) => sum + convertAmount(payment.amount, payment.currency || "USDT", targetCurrency || meta.currency || "USDT", meta), 0);
}

export function totalPayments(payments = []) {
  return payments.reduce((sum, payment) => sum + Number(payment.amount || 0), 0);
}

export function useMoney() {
  const { meta } = useAuth();
  const { locale } = useLocale();

  return {
    convertToEur: (amount, currency) => convertToEur(amount, currency, meta),
    convertAmount: (amount, from, to) => convertAmount(amount, from, to, meta),
    usdRubRate: () => usdRubRate(meta),
    formatMoney: (value, currency = "USDT") => formatMoney(value, currency, locale),
    paymentsTotalIn: (payments = [], targetCurrency) => paymentsTotalIn(payments, targetCurrency, meta),
    formatPaymentTotal: (payments = [], currency) => formatMoney(paymentsTotalIn(payments, currency || meta.currency, meta), currency || meta.currency, locale),
    totalPayments,
  };
}
